engine/src/sellst.cpp

```diff
@@ -98,87 +111,58 @@ bool MCSellist::getids(MCListRef& r_list)
 	if (!MCListCreateMutable('\n', &t_list))
 		return false;
 
-	if (objects != nil)
-	{
-		MCSelnode *tptr = objects;
-		do
-		{
-			MCAutoValueRef t_string;
-            if (tptr->m_ref)
-            {
-                if (!tptr->m_ref->names(P_LONG_ID, &t_string))
-                    return false;
-                if (!MCListAppend(*t_list, *t_string))
-                    return false;
-            }
-			tptr = tptr->next();
-		}
-		while (tptr != objects);
-	}
+    for (const auto& t_node : m_objects)
+    {
+        if (t_node)
+        {
+            MCAutoValueRef t_string;
+            if (!t_node->names(P_LONG_ID, &t_string))
+                return false;
+            if (!MCListAppend(*t_list, *t_string))
+                return false;
+        }
+    }
 
 	return MCListCopy(*t_list, r_list);
 }
 
 void MCSellist::Clean()
 {
-    if (objects == nil)
-        return;
-    
-    // Remove any dead objects from the selected list
-    MCSelnode* t_cursor = objects;
-    do
-    {
-        if (!t_cursor->m_ref)
-        {
-            MCSelnode* t_next = t_cursor->next();
-            t_cursor->remove(objects);
-            delete t_cursor;
-            t_cursor = t_next;
-        }
-        else
-        {
-            t_cursor = t_cursor->next();
-        }
-    }
-    while (t_cursor != objects);
+    /* Remove any dead objects from the selected list. */
+    m_objects.erase(
+         std::remove_if(m_objects.begin(), m_objects.end(),
+                        [](const MCSelnode& node) -> bool { return !node; }),
+         m_objects.end());
 }
 
 void MCSellist::clear(Boolean message)
 {
 	if (locked)
 		return;
     
-	MCObjectHandle optr = nil;
-	while (objects != nil)
-	{
-		MCSelnode *nodeptr = objects->remove(objects);
-		optr = nodeptr->m_ref;
-		delete nodeptr;
-	}
+    const MCObjectHandle optr = [this]() -> MCObjectHandle {
```

Comment:
Looking at the original code here then it looks like optr is actually just used as a boolean flag for 'if there was an object' - i.e. clearing removed at least one. So could this be simplified to:
```
bool t_changed = !m_objects.empty();
m_objects.clear();
if (message && t_changed)
  ...
```